File: day19_part2_2020.php

<?php

$file = file_get_contents("input.txt");
$input = trim($file);
$result = solve($input);
echo $result . PHP_EOL;

function solve($input) {
    list($graph, $messages) = parseInput($input);

    fillInGraph($graph, 42);
    fillInGraph($graph, 31);

    $part42 = "(" . implode("|", $graph[42]->resolved) . ")";
    $part31 = "(" . implode("|", $graph[31]->resolved) . ")";

    $rule8String = "(" . $part42 . ")+";

    $matchRuleZero = 0;
    foreach ($messages as $m) {
        for ($i = 1; $i < 10; $i++) {
            $pattern = "/^" . $rule8String . $part42 . "{" . $i . "}" . $part31 . "{" . $i . "}$/";
            if (preg_match($pattern, $m)) {
                $matchRuleZero++;
                break;
            }
        }
    }

    return $matchRuleZero;
}

function fillInGraph(&$graph, $entry) {
    if (!empty($graph[$entry]->resolved)) {
        return $graph[$entry]->resolved;
    }

    foreach ($graph[$entry]->options as $option) {
        $resolved = [""];
        foreach ($option as $entryPoint) {
            $nestedResolveVals = fillInGraph($graph, $entryPoint);
            $newResolved = [];
            foreach ($nestedResolveVals as $nextPiece) {
                foreach ($resolved as $prev) {
                    $newResolved[] = $prev . $nextPiece;
                }
            }
            $resolved = $newResolved;
        }
        $graph[$entry]->resolved = array_merge($graph[$entry]->resolved, $resolved);
    }

    return $graph[$entry]->resolved;
}

class Rule {
    public $resolved = [];
    public $options = [];
}

function parseInput($input) {
    $parts = explode("\n\n", $input);

    $rules = [];
    foreach (explode("\n", $parts[0]) as $r) {
        if (preg_match("/[a-z]/", $r)) {
            list($num, $char) = sscanf($r, "%d: \"%1s\"");
            $rules[$num] = new Rule();
            $rules[$num]->resolved = [$char];
        } else {
            list($key, $rule) = explode(": ", $r);
            $key = intval($key);
            $newRule = new Rule();
            foreach (explode(" | ", $rule) as $ruleNums) {
                $nums = explode(" ", $ruleNums);
                $option = [];
                foreach ($nums as $n) {
                    $option[] = intval($n);
                }
                $newRule->options[] = $option;
            }
            $rules[$key] = $newRule;
        }
    }

    $messages = explode("\n", $parts[1]);

    return [$rules, $messages];
}

File: dart/day19_part2_2020.dart

import 'dart:io';
import 'dart:convert';

void main() {
  final input = File('input.txt').readAsStringSync().trim();
  final result = solve(input);
  print(result);
}

int solve(String input) {
  final parts = input.split('\n\n');
  final rules = <int, List<List<int>>>{};
  final messages = parts[1].split('\n');

  for (final r in parts[0].split('\n')) {
    if (r.contains(RegExp(r'[a-z]'))) {
      final match = RegExp(r'(\d+): "(\w)"').firstMatch(r)!;
      rules[int.parse(match.group(1)!)] = [[int.parse(match.group(2)!.codeUnitAt(0).toString())]];
    } else {
      final match = RegExp(r'(\d+): (.+)').firstMatch(r)!;
      final key = int.parse(match.group(1)!);
      rules[key] = match.group(2)!.split(' | ').map((e) => e.split(' ').map(int.parse).toList()).toList();
    }
  }

  final resolved42 = resolve(rules, 42).toSet();
  final resolved31 = resolve(rules, 31).toSet();

  final part42 = '(${resolved42.join('|')})';
  final part31 = '(${resolved31.join('|')})';
  final rule8String = '(${part42})+';

  int matchRuleZero = 0;
  for (final m in messages) {
    for (int i = 1; i < 10; i++) {
      final pattern = RegExp('^$rule8String${part42}{${i}}${part31}{${i}}\$');
      if (pattern.hasMatch(m)) {
        matchRuleZero++;
        break;
      }
    }
  }

  return matchRuleZero;
}

Set<String> resolve(Map<int, List<List<int>>> rules, int entry) {
  final resolved = <String>{};
  if (rules[entry]!.first.first < 32) {
    resolved.add(String.fromCharCode(rules[entry]!.first.first));
    return resolved;
  }

  for (final option in rules[entry]!) {
    final nestedResolveVals = option.map((e) => resolve(rules, e)).toList();
    final newResolved = <String>{};
    for (final nextPiece in nestedResolveVals[0]) {
      if (nestedResolveVals.length == 1) {
        newResolved.add(nextPiece);
      } else {
        for (final prev in nestedResolveVals[1]) {
          newResolved.add('$nextPiece$prev');
        }
      }
    }
    resolved.addAll(newResolved);
  }

  return resolved;
}
